forte/data/extractor/base_extractor.py

```diff
@@ -0,0 +1,244 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from abc import ABC
+import logging
+from typing import Tuple, Set, List, Dict, Any
+from typing import Union, Type, Hashable, Iterable, Optional
+from ft.onto.base_ontology import Annotation
+from forte.common.configuration import Config
+from forte.data.data_pack import DataPack
+from forte.data.vocabulary import Vocabulary
+from forte.data.converter.feature import Feature
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "BaseExtractor"
+]
+
+
+class BaseExtractor(ABC):
+    r"""The functionality of Extractor is as followed,
+        1. Build vocabulary.
+        2. Extract feature from datapack.
+        3. Add prediction back to datapack.
```

Comment:
It would be nice to explain why these functionalities are related and should be presented in the same class.

In short, 

Vocabulary is to control a mapping from data -> index,
Feature is represented base on these indices,
And when adding prediction back we need to do index -> data.

Is that correct?

Reply:
Basically, you mean making the explanation more detailed, right? I will do it.
Yes, what you said is correct.

Reply:
As for the previous discussion on the `removal` problem. I am think, maybe we can add one more abstract function in the base extract like `remove_from_pack` opposite to `add_to_pack`. Because I thinking different extractors might have different things to remove. So instead of make another processor to remove the data ( like we might need to have different type of removal processor like extractor, attribute remover, entry remover), it might be more easier to make this as another function in the extractor?
I think I am going to implement it as a function under the extractor. Does that seem good?

Reply:
I see. Since the input to the function is DataPack, I think users can still implement this as a pipeline (like using the ones I suggested). 

Probably we don't need to functions `remove` and `add`, all we need to do is some pre-extract clean-up right? So we can just put one function there to give the user a chance to cleanup. He can cleanup whatever he wants.

This function also doesn't have to be abstract. If the user don't want to clean up, he skips.

Reply:
For attribute extractor, the user would probably want to set the attribute to None or some default value. For seqTaggingExtractor, the user would probably remove the entry, which is EntityMention itself. So I think making it as a function under the extractor is a good design. And we can make the `train_preprocessor` call it during the evaluation, for other time, it is not used unless the users want to.

Reply:
Sure, we can make a general function here, and make the default implementation to be empty.